### PR TITLE
Rewriting enumerable methods

### PR DESCRIPTION
**Pull request by Jmagero**
In this project I have  implemented my own version of Ruby's Enumerables, which include each, each_with_index. any?, all?, none? count, map and inject.

### CODE REVIEW

**Review by MarioDena — CHANGES_REQUESTED** [review]
Hello! Good job on your project however there are some changes that need to be made. :smiley: 

## Changes required 

- [ ] my_each should returns the array itself when a block is given after calling the given block once for each element in self. Please implement this on your method
- [ ] my_each_with_index Same as above.
- [ ] my_none? method is missing please implement it
- [ ] my_count should counts the number of items in enum that are equal to item if an argument is given, please implement this functionality
- [ ] my_map method is missing please implement it
- [ ] my_inject should, when a symbol is specified combines each element of the collection by applying the symbol as a named method.
- [ ] There is something wrong with your stickler set-up and no linter test are shown on your PR. please fix this.

After making the requested changes please re-submit your code for review! Cheers! :+1:

**Comment by Jmagero** [comment]
### TSE required changes fixed

-  my_each should returns the array itself when a block is given after calling the given block once for each element in self. Please implement this on your method
- my_each_with_index Same as above.
- my_none? method is missing please implement it
- my_count should counts the number of items in enum that are equal to item if an argument is given, please implement this functionality
- my_map method is missing please implement it
- my_inject should, when a symbol is specified combines each element of the collection by applying the symbol as a named method.
- There is something wrong with your stickler set-up and no linter test are shown on your PR. please fix this.

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    return enum_for unless block_given?
+
+    my_arr = *self
+    i = 0
+    while i < my_arr.length
+      yield my_arr[i]
+      i += 1
+    end
+    my_arr
+  end
```
Comment (Torres-ssf) [inline main.rb]: - [ ] my_each when a block is given when self is a range after calling the given block once for each element in self, returns the range itself. Your method is returning an array, even when calling my_each from a range

Example:
`(1..3).my_each { | value | puts value }` should print 1, 2 and 3, and after that should return the range itself `1..3`

Print from the original method:

Screenshot from 2020-06-05 09-19-36

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    return enum_for unless block_given?
+
+    my_arr = *self
+    i = 0
+    while i < my_arr.length
+      yield my_arr[i], i
+      i += 1
+    end
+    my_arr
+  end
```
Comment (Torres-ssf) [inline main.rb]: - [ ] my_each_with_index when a block is given when self is a range after calling the given block once for each element in self, returns the range itself. Your method is returning an array, even when calling my_each_with_index from a range

Example:
`(1..3).my_each_with_index { | value, index | puts index }` should print 0, 1 and 2, and after that should return the range itself `1..3`

Print from the original method:

Screenshot from 2020-06-05 09-22-33

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+      elsif pattern.class == Class
+        return false unless x.class == pattern
+      elsif !pattern.nil?
+        return false unless x == pattern
+      else
+        return false unless x
+      end
+    end
+    true
+  end
```
Comment (Torres-ssf) [inline main.rb]: - [ ] my_all? when a class is passed as an argument returns true if all of the collection is a member of such Class. 
Example:

`[1, 2.14, 0.5, 0].my_all?(Numeric)` should return true
`[1, 2.14, 0.5, 0].my_all?(Integer)` should return false
`[1, 2.14, 0.5, 'hello'].my_all?(String)` should return false

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+      elsif !pattern.nil?
+        return true if x == pattern
+      else
+        return true unless x
+      end
+    end
+    return true if pattern.nil? && !block_given?
+
+    false
+  end
```
Comment (Torres-ssf) [inline main.rb]: - [ ] my_any? returns false if the block ever returns a value false or nil. This method is working OK for an array, but it throws an error when calling it from a range.

Example:

`(1..10).my_any? { | value | value > 10 }` should return false
`(1..11).my_any? { | value | value > 10 }` should return true

Print from the original method:
Screenshot from 2020-06-05 09-44-27


- [ ] my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class.

`[1, 2.14, 0.5, 0].my_any?(Integer)` should return true
`[2.14, 0.5].my_any?(Integer)` should return false
`[1, 2.14, 0.5, 'hello'].my_any?(String)` should return true

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+      elsif pattern.class == Class
+        return false if x.class == pattern
+      elsif !pattern.nil?
+        return false if x == pattern
+      elsif i.positive? && self[i] != self[i - 1]
+        return false
+      end
+    end
+    true
+  end
```
Comment (Torres-ssf) [inline main.rb]: - [ ] my_none? returns false if the block ever returns true for all elements. This method is working OK for an array, but it throws an error when calling it from a range.

Example:

`(1..10).my_none? { | value | value > 11 }` should return true
`(1..10).my_none? { | value | value > 9 }` should return false

Print from the original method:
Screenshot from 2020-06-05 09-52-04

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    my_each do |x|
+      if block_given?
+        count += 1 if yield x
+      elsif x == arg
+        count += 1
+      end
+    end
+
+    count
+  end
```
Comment (Torres-ssf) [inline main.rb]: - [ ] my_count returns the number of items in enum through enumeration. This method is working OK for an array, but it throws an error when calling it from a range.

`(1..10).my_count` should return 10

Print from the original method:
Screenshot from 2020-06-05 09-57-19

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+      elsif start == :/
+        result = arr.my_inject { |i, v| i / v }
+      end
+    elsif start.class == Integer && arg.class == Symbol
+      new_arr = arr.to_a
+      new_arr.unshift(start)
+      result = new_arr.my_inject(arg)
+    end
+    result
+  end
```
Comment (Torres-ssf) [inline main.rb]: - [ ] my_inject raises a "LocalJumpError" when no block or argument is given

Example:

`[1, 2, 3, 4, 5].inject` should raise a LocalJumpError

Print original method:
Screenshot from 2020-06-05 10-02-24

- [ ] my_inject, when a block is given without an initial value, combines all elements of enum by applying a binary operation, specified by a block

Example: 

`[1, 2, 3, 4, 5].my_inject { |sum, value| sum + value }` should return 15
`[1, 2, 3, 4, 5].my_inject { |product, value| product * value }` should return 120
`[1, 2, 3, 4, 5].my_inject { |sub, value| sub - value }` should return 13

Should also work for Range:

`(1..5).inject { |sum, value| sum + value }` should return 15

**Review by Torres-ssf — CHANGES_REQUESTED** [review]
### Changes required :muscle: 
Hi there @Jmagero 
Great working on the requested changes. However, there still some methods that we need to improve. The reason for this is because in this assignment your methods should work very similar to the original ones. **Check my comments bellow**

Please kindly submit for code review again after making these changes.
Happy coding
**Sergio Torres**

**Comment by Jmagero** [comment]
### Fixed these errors
-  my_each when a block is given when self is a range after calling the given block once for each element in self, returns the range itself.
- my_each_with_index when a block is given when self is a range after calling the given block once for each element in self, returns the range itself. 
- my_all? when a class is passed as an argument returns true if all of the collection is a member of such Class.
- my_any? returns false if the block ever returns a value false or nil. This method is working OK for an array, but it throws an error when calling it from a range.
- my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class.
- my_none? returns false if the block ever returns true for all elements. This method is working OK for an array, but it throws an error when calling it from a range.
- my_count returns the number of items in enum through enumeration. This method is working OK for an array, but it throws an error when calling it from a range.
- my_inject raises a "LocalJumpError" when no block or argument is given

File: README.md [hunk 8]
```diff
@@ -1 +1,39 @@
-# Enumerable-methods
\ No newline at end of file
+# project-bubble-sort
```
Comment (euqueme) [inline README.md]: - [ ] The project is called Enumerable-methods not bubble-sort let's correct it

File: main.rb [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+      if block_given?
+        return true if yield(x)
+      elsif args.nil?
+        return true if x === args
+      elsif x
+        return true
+      end
+    end
+    false
+  end
```
Comment (euqueme) [inline main.rb]: - [ ] should have working `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

-  When no block or argument is given returns false if at least one of the collection is not true
```ruby
[false, false, nil].my_any? == [false, false, nil].any? # => false
```
Yours (wrong answer)
image

Real any? (correct)
image

- When a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer

```Ruby
['1', '2', '5', false].my_any?(Integer) == ['1', '2', '5', false].any?(Integer) # => false
```
Yours (wrong answer)
image

Real any? (correct)
image

- when a Regex is passed as an argument returns false if none of the collection matches the Regex

```ruby
['lorem','ipsum', 'dolo'].my_any?(/z/) == ['lorem','ipsum', 'dolo'].any?(/z/) # => false
```
Yours (wrong answer)
image

Real any? (correct)
image

- When a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
```ruby
['lorem','ipsum', 'dolo'].my_any?('cat') == ['lorem','ipsum', 'dolo'].any?('cat') # => false
```
Yours (wrong answer)
image

Real any? (correct)
image

File: main.rb [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+      elsif start == :/
+        result = arr.my_inject { |i, v| i / v }
+      end
+    elsif start.class == Integer && arg.class == Symbol
+      new_arr = arr.to_a
+      new_arr.unshift(start)
+      result = new_arr.my_inject(arg)
+    end
+    result
+  end
```
Comment (euqueme) [inline main.rb]: - [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

- raises a "LocalJumpError" when no block or argument is given
```ruby
[1, 5, 8, 10].my_inject == [1, 5, 8, 10].inject 
```
Yours (incorrect)
image

Real inject (correct)
image

- It searches for the element that meets the condition within a proc given as an argument

```ruby
search = proc { |memo, word| memo.length > word.length ? memo : word }
['lorem','ipsum', 'dolo'].my_inject(&search) == ['lorem','ipsum', 'dolo'].inject(&search) # => "ipsum"
```
Yours (incorrect)
image

Real inject (correct)
image

- when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block

```ruby
operation = proc { |sum, n| sum + n } 
[5, 10, 15, 20].my_inject(&operation) == [5, 10, 15, 20].inject(&operation) # => 50
```
Yours (Incorrect)
image

Real inject (correct)
image

- when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::Range

```ruby
(1..10).my_inject { |prod, n| prod * n } == (1..10).my_inject { |prod, n| prod * n } # => 3628800
```
Yours (Incorrect)
image

Real inject (correct)
image

**Review by euqueme — CHANGES_REQUESTED** [review]
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @Jmagero :balloon: 

### Thank you for doing the changes requested by the previous reviewers :medal_sports: your project is working better and better every time, :muscle: please address to the changes I've left in your code :point_down: 

### That's it! :raised_hands: 

### Please :pray:  do the changes :woman_technologist: and submit back for a code review when you are done :memo: 

Happy coding! :rocket: 

Any questions? ask me on slack **@ Maru**

PROJECT STATUS: REQUESTED CHANGES :vertical_traffic_light:

**Comment by Jmagero** [comment]
-  The project is called Enumerable-methods not bubble-sort let's correct it -Fixed

**Review by kobiyoyo — CHANGES_REQUESTED** [review]
## Hi @Jmagero         
#### Great Job so far, but there are a few more things required. 
### General

- [ ] `my_any?`:
- when no block or argument is given returns true if at least one of the collection is not false or nil
`p [nil, false, nil, false].my_any?` - should return `false`
- when a class is passed as an argument returns true if at least one of the collection is a member of such class
`p [12,3,34].my_any?(String)` should return `false`
- when a Regex is passed as an argument returns false if none of the collection matches the Regex
`p ['ghetto','wild'].my_any?(/z/)` should return `false`
- when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
`p ['cat',2,32,].my_any?('at')` should return `false`
- [ ] `my_inject`:
- When a symbol is specified combines each element of the collection by applying the symbol as a named method
`p (5..10).my_inject(2, :*) - should return 302400`
- Combines all elements of enum by applying a binary operation, specified by a block:
 `p (1..5).my_inject(4) { |prod, n| prod * n }` should return 480

### Make changes and submit another code review request.
*Happy coding* 
(Any questions Feel free to contact me on slack @kobiyoyo)

**Comment by Jmagero** [comment]
### Fixed
- my_any?:
when no block or argument is given returns true if at least one of the collection is not false or nil
p [nil, false, nil, false].my_any? - should return false
when a class is passed as an argument returns true if at least one of the collection is a member of such class
p [12,3,34].my_any?(String) should return false
when a Regex is passed as an argument returns false if none of the collection matches the Regex
p ['ghetto','wild'].my_any?(/z/) should return false
when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
p ['cat',2,32,].my_any?('at') should return false
- my_inject:
When a symbol is specified combines each element of the collection by applying the symbol as a named method
p (5..10).my_inject(2, :*) - should return 302400
Combines all elements of enum by applying a binary operation, specified by a block:
p (1..5).my_inject(4) { |prod, n| prod * n } should return 480

**Review by Torres-ssf — APPROVED** [review]
### Project Approved :clap: :clap: :clap: 
Hi there @Jmagero 
Great work on this project :+1: 
I'm approving this PR :smile: 

**Only one thing to comment though, the `multiply_els` should not belong to the Enumerable module. So you can just remove it from the module and place below it.**

Good luck in the next Project

Happy coding :computer: 
**Sergio Torres**

**Pull request merged**